### PR TITLE
changed repository and created search method

### PR DESCRIPTION
**Pull request by JeniaSan**

### CODE REVIEW

File: src/main/java/com/polishuchenko/bookstore/repository/book/specification/PriceSpecificationProvider.java [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+import org.springframework.data.jpa.domain.Specification;
+import org.springframework.stereotype.Component;
+
+@Component
+public class PriceSpecificationProvider implements SpecificationProvider<Book> {
+    private static final String KEY = "price";
+
+    @Override
+    public Specification<Book> getSpecification(String[] params) {
+        if (params.length <= 1) {
```
Comment (Someboty) [inline src/main/java/com/polishuchenko/bookstore/repository/book/specification/PriceSpecificationProvider.java]: if params.length == 0, you will call that block too?
```suggestion
        if (params.length == 1) {
```

File: src/main/java/com/polishuchenko/bookstore/repository/book/specification/PriceSpecificationProvider.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+            return (root, query, criteriaBuilder) -> criteriaBuilder.lessThanOrEqualTo(
+                    root.get(KEY), Arrays.stream(params)
+                            .map(Integer::parseInt)
+                            .max(Comparator.comparingInt(Integer::intValue))
+                            .orElse(Integer.MAX_VALUE));
```
Comment (Someboty) [inline src/main/java/com/polishuchenko/bookstore/repository/book/specification/PriceSpecificationProvider.java]: Does “Arrays…” part necessary?

File: src/main/java/com/polishuchenko/bookstore/repository/book/specification/PriceSpecificationProvider.java [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+        }
+        return (root, query, criteriaBuilder) -> criteriaBuilder.and(
+            criteriaBuilder.greaterThanOrEqualTo(root.get(KEY), Arrays.stream(params)
+                    .map(Integer::parseInt)
+                    .min(Comparator.comparingInt(Integer::intValue))
+                    .get()),
+                criteriaBuilder.lessThanOrEqualTo(root.get(KEY), Arrays.stream(params)
+                        .map(Integer::parseInt)
+                        .max(Comparator.comparingInt(Integer::intValue))
+                        .get()));
```
Comment (Someboty) [inline src/main/java/com/polishuchenko/bookstore/repository/book/specification/PriceSpecificationProvider.java]: I think you can simplify it with criteriaBuilder.between()

File: src/main/resources/db/changelog/db.changelog-master.yaml [hunk 4]
```diff
@@ -0,0 +1,3 @@
+databaseChangeLog:
+  - include:
+      file: db/changelog/changes/create-book-table.yaml
```
Comment (Someboty) [inline src/main/resources/db/changelog/db.changelog-master.yaml]: You forgot empty line here)

**Review by Someboty — COMMENTED** [review]
Great job! Let's make it a little better)

File: src/main/java/com/polishuchenko/bookstore/repository/book/specification/PriceSpecificationProvider.java (hunk 3 above)
Comment (JeniaSan) [reply src/main/java/com/polishuchenko/bookstore/repository/book/specification/PriceSpecificationProvider.java]: i tried this, but this option is not working (don't know why)

File: src/main/java/com/polishuchenko/bookstore/repository/book/BookSpecificationBuilder.java [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+    private static final String AUTHOR_KEY = "author";
+    private static final String ISBN_KEY = "isbn";
+    private static final String PRICE_KEY = "price";
+    private static final String TITLE_KEY = "title";
+    private final SpecificationProviderManager<Book> specificationProviderManager;
+
+    @Override
+    public Specification<Book> build(BookSearchParameters searchParameters) {
+        Specification<Book> specification = Specification.where(null);
+        if (searchParameters.author() != null && searchParameters.author().length > 0) {
```
Comment (vovasalyha) [inline src/main/java/com/polishuchenko/bookstore/repository/book/BookSpecificationBuilder.java:23]: You are writing the same code `searchParameters.author() != null && searchParameters.author().length > 0` many times, for me it's a good candidate for some private method.

File: src/main/java/com/polishuchenko/bookstore/repository/book/BookSpecificationProviderManager.java [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+@Component
+public class BookSpecificationProviderManager implements SpecificationProviderManager<Book> {
+    private final List<SpecificationProvider<Book>> bookSpecificationProviders;
+
+    @Override
+    public SpecificationProvider<Book> getSpecificationProvider(String key) {
+        return bookSpecificationProviders.stream()
+                .filter(p -> p.getKey().equals(key))
+                .findFirst()
+                .orElseThrow(() -> new RuntimeException(
```
Comment (vovasalyha) [inline src/main/java/com/polishuchenko/bookstore/repository/book/BookSpecificationProviderManager.java:19]: better throw `NoSuchElement`

**Review by vovasalyha — APPROVED** [review]
GJ. Left few suggestions.

File: src/main/java/com/polishuchenko/bookstore/repository/book/BookSpecificationBuilder.java (hunk 5 above)
Comment (andrii-hoienko) [reply src/main/java/com/polishuchenko/bookstore/repository/book/BookSpecificationBuilder.java:23]: Please use already existing method for example StringUtils.isEmpty() or create a custom one

File: src/main/java/com/polishuchenko/bookstore/repository/book/BookSpecificationBuilder.java [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+@Component
+public class BookSpecificationBuilder implements SpecificationBuilder<Book, BookSearchParameters> {
+    private static final String AUTHOR_KEY = "author";
+    private static final String ISBN_KEY = "isbn";
+    private static final String PRICE_KEY = "price";
+    private static final String TITLE_KEY = "title";
+    private final SpecificationProviderManager<Book> specificationProviderManager;
+
+    @Override
+    public Specification<Book> build(BookSearchParameters searchParameters) {
```
Comment (andrii-hoienko) [inline src/main/java/com/polishuchenko/bookstore/repository/book/BookSpecificationBuilder.java:21]: OPTIONAL: You can implement strategy here, current implementation violates Open-closed principle

File: src/main/java/com/polishuchenko/bookstore/repository/book/BookSpecificationProviderManager.java (hunk 6 above)
Comment (andrii-hoienko) [reply src/main/java/com/polishuchenko/bookstore/repository/book/BookSpecificationProviderManager.java:19]: Agree

**Review by andrii-hoienko — COMMENTED** [review]
GJ! Let`s improve your solution

File: src/main/java/com/polishuchenko/bookstore/repository/book/BookSpecificationBuilder.java (hunk 7 above)
Comment (JeniaSan) [reply src/main/java/com/polishuchenko/bookstore/repository/book/BookSpecificationBuilder.java:21]: All non-optional changes are made, but they were already done in another branch. Sorry for the confusion. Check this changes in this branch https://github.com/JeniaSan/Online-Book-Store/pull/4/files#diff-35f1cce186d0f9006029185eab2efc2c731bd15fdd7a2e0a058ac0964408f923

**Pull request merged**